File: scripts/canonical_hash.py
#!/usr/bin/env python3
"""
canonical_hash.py — harness-boot canonical hashing (F-010)

사용:
  python3 scripts/canonical_hash.py <path/to/spec.yaml>           # root hash 만
  python3 scripts/canonical_hash.py <path/to/spec.yaml> --subtrees # root + subtree 맵
  python3 scripts/canonical_hash.py <path/to/spec.yaml> --json    # JSON 출력 (harness.yaml 에 merge 용이)

알고리즘 (부록 D 기반):
  1. YAML 을 Python 객체로 파싱 (주석·앵커·순서 무시).
  2. 객체를 **canonical JSON** 으로 직렬화 — sort_keys=True, separators=(',', ':').
     Unicode 는 그대로 (ensure_ascii=False) — 한국어 스펙 필드 해시 안정성.
  3. UTF-8 인코딩 후 SHA-256 → 16진수 64자.

Subtree 해시:
  project · domain · constraints · deliverable · features · metadata
  각각 독립 해시. features 는 배열이라 그대로 해시 (순서 영향 O).

Merkle 결합:
  subtree 해시들을 sorted(key: hash) 배열로 묶어 canonical JSON → SHA-256
  = merkle_hash. Full spec hash 와 다름 (다른 필드 주입 감지 가능).

제약:
  - $include 전개는 이 스크립트 범위 외. 입력은 이미 전개된 spec 을 가정.
  - 두 스펙이 **구조 동등** 하면 (YAML 표기만 다름) 같은 hash. 의도된 동작.
  - 외부 의존: PyYAML (pyyaml). ruamel.yaml 불필요 (canonical 은 순서 무시).

참조:
  - design/harness-boot-design-2.3.7.md 부록 D (원본 명세, gitignore 내부 문서).
  - docs/samples/harness-boot-self/spec.yaml F-010 — acceptance_criteria.
"""

from __future__ import annotations

import argparse
import hashlib
import json
import sys
from pathlib import Path
from typing import Any

try:
    import yaml
except ImportError:
    print("pyyaml is required. Install: pip install pyyaml", file=sys.stderr)
    sys.exit(1)


def canonical_bytes(obj: Any) -> bytes:
    """객체 → canonical JSON bytes (UTF-8)."""
    return json.dumps(
        obj,
        sort_keys=True,
        ensure_ascii=False,
        separators=(",", ":"),
        allow_nan=False,
    ).encode("utf-8")


def canonical_hash(obj: Any) -> str:
    """객체 → SHA-256 16진수."""
    return hashlib.sha256(canonical_bytes(obj)).hexdigest()


# 고정 subtree 키 — 스펙 top-level 에서 해시할 것들.
SUBTREE_KEYS = (
    "project",
    "domain",
    "constraints",
    "deliverable",
    "features",
    "metadata",
)


def subtree_hashes(spec: dict) -> dict:
    """각 subtree 의 canonical hash 맵."""
    result = {}
    for key in SUBTREE_KEYS:
        if key in spec:
            result[key] = canonical_hash(spec[key])
    return result


def merkle_root(subtrees: dict) -> str:
    """subtree 해시들을 결합한 merkle root."""
    # sorted by key → 순서 독립적
    combined = [{"key": k, "hash": v} for k, v in sorted(subtrees.items())]
    return canonical_hash(combined)


def compute_all(spec: dict) -> dict:
    """전체 해시 번들. harness.yaml.generation.generated_from 과 매핑 가능."""
    st = subtree_hashes(spec)
    return {
        "spec_hash": canonical_hash(spec),
        "subtrees": st,
        "merkle_root": merkle_root(st),
    }


def load_spec(path: Path) -> dict:
    with path.open("r", encoding="utf-8") as f:
        data = yaml.safe_load(f)
    if not isinstance(data, dict):
        raise ValueError(f"{path}: top-level YAML 이 mapping 이 아님")
    return data


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Compute canonical hash for a harness-boot spec.yaml",
    )
    parser.add_argument("spec", type=Path, help="path to spec.yaml")
    parser.add_argument(
        "--subtrees",
        action="store_true",
        help="print subtree hashes in addition to root",
    )
    parser.add_argument(
        "--json",
        action="store_true",
        help="emit a JSON object (suitable for piping into jq)",
    )
    args = parser.parse_args(argv)

    if not args.spec.is_file():
        print(f"error: {args.spec} not found", file=sys.stderr)
        return 2

    spec = load_spec(args.spec)
    bundle = compute_all(spec)

    if args.json:
        if not args.subtrees:
            # subtrees 안 보여달라 요청이면 비워서 출력 (merkle 는 유지)
            bundle_out = {k: v for k, v in bundle.items() if k != "subtrees"}
        else:
            bundle_out = bundle
        json.dump(bundle_out, sys.stdout, indent=2, ensure_ascii=False)
        print()  # trailing newline
    else:
        print(f"spec_hash   {bundle['spec_hash']}")
        print(f"merkle_root {bundle['merkle_root']}")
        if args.subtrees:
            print("subtrees:")
            for k, v in bundle["subtrees"].items():
                print(f"  {k:12} {v}")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/core/canonicalHash.ts
/**
 * Canonical hashing for harness-boot specs (F-010, ported in F-084).
 *
 * This module is the single most critical port of the Python → TypeScript
 * migration: every drift detector and every `harness.yaml.spec_hash`
 * record depends on the hash being identical across runtimes. A
 * one-byte deviation breaks the entire drift system.
 *
 * Mirrors `scripts/core/canonical_hash.py` byte-for-byte:
 *
 *   1. Parse YAML to a plain object (comments / anchors / order
 *      stripped). Performed by the caller; this module accepts a
 *      pre-parsed value.
 *   2. Serialize via canonical JSON — keys sorted recursively, no
 *      whitespace separators, Unicode preserved, NaN / Infinity
 *      rejected.
 *   3. UTF-8 encode and SHA-256 → 64-char lowercase hex.
 *
 * Equivalent Python invocation:
 *
 *   json.dumps(obj, sort_keys=True, ensure_ascii=False,
 *              separators=(",", ":"), allow_nan=False)
 *
 * Native `JSON.stringify` does not sort keys and does not raise on
 * NaN / Infinity (it converts to `null`), so this module canonicalizes
 * the input first and validates non-finite numbers explicitly before
 * passing the result to `JSON.stringify`.
 *
 * @module canonicalHash
 */

import {createHash} from 'node:crypto';

/**
 * Subtree keys hashed independently for the spec's Merkle tree.
 *
 * Mirrors `scripts/core/canonical_hash.py:SUBTREE_KEYS`. Order is
 * informational only — the Merkle root sorts entries by key when
 * combining hashes.
 */
export const SUBTREE_KEYS = [
  'project',
  'domain',
  'constraints',
  'deliverable',
  'features',
  'metadata',
] as const;

export type SubtreeKey = (typeof SUBTREE_KEYS)[number];

/**
 * Bundle of every hash the harness-boot spec needs.
 *
 * Mirrors the dict returned by
 * `scripts/core/canonical_hash.py:compute_all`:
 *
 *   - `specHash`: hash of the entire parsed spec, top-level keys sorted.
 *   - `subtrees`: per-subtree hash, only keys that are present in the
 *     input appear (e.g. `metadata` is omitted when the spec has none).
 *   - `merkleRoot`: hash of `[{key, hash}, ...]` sorted by `key`. Order-
 *     independent so reordering subtree keys in the input does not
 *     change the root.
 */
export interface HashBundle {
  specHash: string;
  subtrees: Partial<Record<SubtreeKey, string>>;
  merkleRoot: string;
}

/**
 * Recursively rebuilds the input value with object keys sorted.
 *
 * Arrays preserve order (sequence is part of the canonical form).
 * Plain objects produce a new object whose keys appear in
 * lexicographically sorted order — `JSON.stringify` then walks them in
 * insertion order, matching Python's `sort_keys=True` behaviour.
 *
 * Non-finite numbers (`NaN`, `Infinity`, `-Infinity`) throw because
 * Python's `allow_nan=False` does the same. Functions, symbols, and
 * `undefined` likewise throw — they cannot survive a JSON round-trip.
 *
 * @param value - Any JSON-compatible value; objects must be plain.
 * @returns A structurally equivalent value with keys sorted.
 * @throws {TypeError} On non-finite numbers or non-JSON-compatible types.
 */
function canonicalize(value: unknown): unknown {
  if (value === null) {
    return null;
  }
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) {
      throw new TypeError(
        `canonical_hash: non-finite number encountered (${String(value)}); ` +
          'Python equivalent uses allow_nan=False which raises ValueError.',
      );
    }
    return value;
  }
  if (typeof value === 'string' || typeof value === 'boolean') {
    return value;
  }
  if (Array.isArray(value)) {
    return value.map((item) => canonicalize(item));
  }
  if (typeof value === 'object') {
    // Plain object — sort keys recursively. Buffers, Dates, Maps, Sets,
    // and class instances are not canonicalizable and would silently
    // round-trip through JSON.stringify with surprising results.
    const proto = Object.getPrototypeOf(value);
    if (proto !== Object.prototype && proto !== null) {
      throw new TypeError(
        'canonical_hash: only plain objects are canonicalizable; ' +
          `received instance of ${value.constructor?.name ?? 'unknown'}.`,
      );
    }
    const obj = value as Record<string, unknown>;
    const sortedKeys = Object.keys(obj).sort();
    const result: Record<string, unknown> = {};
    for (const key of sortedKeys) {
      result[key] = canonicalize(obj[key]);
    }
    return result;
  }
  throw new TypeError(
    `canonical_hash: cannot serialize value of type ${typeof value}.`,
  );
}

/**
 * Serializes a value to canonical JSON bytes (UTF-8).
 *
 * Mirrors `scripts/core/canonical_hash.py:canonical_bytes`. The output
 * matches Python's
 * `json.dumps(obj, sort_keys=True, ensure_ascii=False,
 * separators=(",", ":"), allow_nan=False).encode("utf-8")` byte-for-byte.
 *
 * @param value - Pre-parsed YAML / JSON value.
 * @returns UTF-8 encoded canonical JSON.
 * @throws {TypeError} When the value contains non-finite numbers or
 *   non-JSON-compatible types.
 */
export function canonicalBytes(value: unknown): Buffer {
  const sorted = canonicalize(value);
  // JSON.stringify with no spacer argument produces no whitespace
  // between separators, matching Python's `(",", ":")` configuration.
  // It also preserves Unicode characters as-is, matching Python's
  // `ensure_ascii=False`.
  const json = JSON.stringify(sorted);
  return Buffer.from(json, 'utf-8');
}

/**
 * Computes the canonical SHA-256 hash of a parsed YAML / JSON value.
 *
 * Mirrors `scripts/core/canonical_hash.py:canonical_hash`. The output
 * is a 64-character lowercase hex string — the same format
 * `harness.yaml.generation.generated_from.spec_hash` expects.
 *
 * @param value - Pre-parsed value (typically from `yaml.parse`).
 * @returns 64-character lowercase hex SHA-256 digest.
 * @throws {TypeError} When the value contains non-finite numbers or
 *   non-JSON-compatible types.
 */
export function canonicalHash(value: unknown): string {
  return createHash('sha256').update(canonicalBytes(value)).digest('hex');
}

/**
 * Computes the per-subtree hash map for a spec.
 *
 * Mirrors `scripts/core/canonical_hash.py:subtree_hashes`. Only keys
 * present in `spec` produce an entry; missing keys are omitted (they
 * are not hashed as `null`).
 *
 * @param spec - Top-level parsed spec object.
 * @returns Map from subtree key to canonical SHA-256 hex.
 */
export function subtreeHashes(
  spec: Record<string, unknown>,
): Partial<Record<SubtreeKey, string>> {
  const result: Partial<Record<SubtreeKey, string>> = {};
  for (const key of SUBTREE_KEYS) {
    if (Object.prototype.hasOwnProperty.call(spec, key)) {
      result[key] = canonicalHash(spec[key]);
    }
  }
  return result;
}

/**
 * Computes the Merkle root that combines every subtree hash.
 *
 * Mirrors `scripts/core/canonical_hash.py:merkle_root`. Builds an array
 * of `{key, hash}` records sorted by `key`, then canonical-hashes the
 * array. Adding a new top-level field (e.g. injected `garbage` key)
 * shifts the root, which is how harness-boot detects unexpected spec
 * structure changes.
 *
 * @param subtrees - Per-subtree hash map (output of
 *   {@link subtreeHashes}).
 * @returns 64-character lowercase hex SHA-256 of the combined record.
 */
export function merkleRoot(
  subtrees: Partial<Record<SubtreeKey, string>>,
): string {
  const entries = Object.entries(subtrees) as Array<[string, string]>;
  entries.sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
  const combined = entries.map(([key, hash]) => ({key, hash}));
  return canonicalHash(combined);
}

/**
 * Computes the full hash bundle for a spec.
 *
 * Mirrors `scripts/core/canonical_hash.py:compute_all`. Returned shape
 * matches `harness.yaml.generation.generated_from` — `spec_hash`
 * becomes the entire-spec digest, `subtrees` carries per-section
 * digests, and `merkle_root` carries the order-independent combined
 * digest.
 *
 * @param spec - Top-level parsed spec object.
 * @returns {@link HashBundle} with every hash needed by harness.yaml.
 */
export function computeAll(spec: Record<string, unknown>): HashBundle {
  const subtrees = subtreeHashes(spec);
  return {
    specHash: canonicalHash(spec),
    subtrees,
    merkleRoot: merkleRoot(subtrees),
  };
}
